dvl-a50/dvl.py

```diff
@@ -175,12 +171,72 @@ def set_should_send(self, should_send):
         self.should_send = should_send
         self.save_settings()
 
-    def set_gps_origin(self, lat, lon):
+    @staticmethod
+    def longitude_scale(lat: float):
+        """
+        from https://github.com/ArduPilot/ardupilot/blob/Sub-4.1/libraries/AP_Common/Location.cpp#L325
+        """
+        scale = math.cos(math.radians(lat))
+        return max(scale, 0.01)
+
+    def lat_lng_to_NE_XY_cm (self, lat: float, lon: float) -> List[float]:
+        """
+        From https://github.com/ArduPilot/ardupilot/blob/Sub-4.1/libraries/AP_Common/Location.cpp#L206
+        """
+        x = (lat-self.origin[0]) * LATLON_TO_CM
+        y =  self.longitude_scale((lat + self.origin[0])/2) * LATLON_TO_CM * (lon-self.origin[1])
+        return [x, y]
+
+    def has_origin_set(self) -> bool:
+        try:
+            old_time = self.mav.get_float("/GPS_GLOBAL_ORIGIN/message/time_usec")
+            if old_time == float('nan'):
+                logger.warning("Unable to read current time for GPS_GLOBAL_ORIGIN, using 0")
+        except Exception as e:
+            logger.warning(f"Unable to read current time for GPS_GLOBAL_ORIGIN, using 0: {e}")
+
+        for attempt in range(5):
+            logger.debug(f"Trying to read origin, try # {attempt}")
+            self.mav.request_message(GPS_GLOBAL_ORIGIN_ID)
+            time.sleep(0.5) # make this a timeout?
+            try:
+                new_origin_data = json.loads(self.mav.get("/GPS_GLOBAL_ORIGIN/message"))
+                if new_origin_data["time_usec"] != old_time:
+                    self.origin = [new_origin_data["latitude"]*1e-7, new_origin_data["longitude"]*1e-7]
+                    return True
+                else:
+                    continue # try again
+            except Exception as e:
+                logger.warning(e)
+                return False
+        return False
+
+    def set_current_position(self, lat: float, lon: float):
+        """
+        Sets the EKF origin to lat, lon
+        """
+        # If origin has never been set, set it
+        if not self.has_origin_set():
+            logger.info("Origin was never set, trying to set it.")
+            self.set_gps_origin(lat, lon)
+        else:
+            logger.info("Origin has already been set, sending POSITION_ESTIMATE instead")
+            # if we already have an origin set, send a new position instead
+            x, y = self.lat_lng_to_NE_XY_cm (lat, lon)
```

Comment:
there is a weird space before the parentheses :eyes: 
The project should run a linter or something to catch this kind of stuff

Reply:
yeah, after this get in, I think I'll add black and isort at least. I'll fix this meanwhile